experimental/normalizer/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+        unsafe { char::from_u32_unchecked(self.0 & 0xFFFFFF) }
+    }
+    pub fn ccc(&self) -> CanonicalCombiningClass {
+        CanonicalCombiningClass((self.0 >> 24) as u8)
+    }
+    pub fn set_ccc_from_trie(&mut self, ccc_trie: &CodePointTrie<CanonicalCombiningClass>) {
+        debug_assert_eq!(self.0 >> 24, 0, "This method has already been called!");
+        self.0 |= (ccc_trie.get(self.0).0 as u32) << 24;
+    }
+}
+
+// This function exists as a borrow check helper.
+#[inline(always)]
+fn sort_slice_by_ccc<'data>(
+    slice: &mut [CharacterAndClass],
+    ccc: &CodePointTrie<'data, CanonicalCombiningClass>,
+) {
+    // We don't look up the canonical combining class for starters
+    // of for single combining characters between starters. When
+    // there's more than one combining character between starters,
+    // we look up the canonical combining class for each character
+    // exactly once.
+    if slice.len() < 2 {
+        return;
+    }
+    slice.iter_mut().for_each(|cc| cc.set_ccc_from_trie(ccc));
+    slice.sort_by_key(|cc| cc.ccc());
+}
+
+/// An iterator adaptor that turns an `Iterator` over `char` into
+/// a lazily-decomposed (currently only NFD though NFKD should also
+/// be possible) `char` sequence.
+pub struct Decomposition<'data, I>
+where
+    I: Iterator<Item = char>,
+{
+    delegate: I,
+    buffer: SmallVec<[CharacterAndClass; 10]>, // TODO Figure out good length
+    /// The index of the next item to be read from `buffer`.
+    /// The purpose if this index is to avoid having to move
+    /// the rest upon every read.
+    buffer_pos: usize,
+    pending_unnormalized_starter: Option<char>, // None at end of stream
+    decompositions: &'data CanonicalDecompositionDataV1<'data>,
+    ccc: &'data CodePointTrie<'data, CanonicalCombiningClass>,
+}
+
+impl<'data, I> Decomposition<'data, I>
+where
+    I: Iterator<Item = char>,
+{
+    /// Constructs a decomposing iterator adapter from a delegate
+    /// iterator and references to the necessary data.
+    ///
+    /// Use `DecomposingNormalizer::normalize_iter()` instead unless
+    /// there's a good reason to use this constructor directly.
+    pub fn new(
+        delegate: I,
+        decompositions: &'data CanonicalDecompositionDataV1,
+        ccc: &'data CodePointTrie<'data, CanonicalCombiningClass>,
+    ) -> Self {
+        let mut ret = Decomposition::<I> {
+            delegate,
+            buffer: SmallVec::new(), // Normalized
+            buffer_pos: 0,
+            // Initialize with a placeholder starter in case
+            // the real stream starts with a non-starter.
+            pending_unnormalized_starter: Some('\u{FFFF}'),
+            decompositions,
+            ccc,
+        };
+        let _ = ret.next(); // Remove the U+FFFF placeholder
+        ret
+    }
+}
+
+impl<'data, I> Iterator for Decomposition<'data, I>
+where
+    I: Iterator<Item = char>,
+{
+    type Item = char;
+
+    fn next(&mut self) -> Option<char> {
```

Comment:
Nit: This function needs some work to make it corrupted-data-safe. Can be a follow-up.

Reply:
This is #1892.